crates/starknet-devnet-server/src/api/json_rpc/endpoints.rs

```diff
@@ -71,22 +83,33 @@ impl JsonRpcHandler {
 
     /// starknet_getStateUpdate
     pub async fn get_state_update(&self, block_id: BlockId) -> StrictRpcResult {
+        let starknet = self.api.starknet.read().await;
+
         let state_update =
-            self.api.starknet.read().await.block_state_update(block_id.as_ref()).map_err(
-                |err| match err {
-                    Error::NoBlock => ApiError::BlockNotFound,
-                    unknown_error => ApiError::StarknetDevnetError(unknown_error),
-                },
-            )?;
+            starknet.block_state_update(block_id.as_ref()).map_err(|err| match err {
+                Error::NoBlock => ApiError::BlockNotFound,
+                unknown_error => ApiError::StarknetDevnetError(unknown_error),
+            })?;
 
         let state_diff = state_update.state_diff.into();
 
-        Ok(StarknetResponse::StateUpdate(StateUpdate {
-            block_hash: state_update.block_hash,
-            new_root: state_update.new_root,
-            old_root: state_update.old_root,
-            state_diff,
-        }))
+        // StateUpdate needs to be mapped to PendingStateUpdate response only in blocks_on_demand
+        // mode and when block_id is pending
+        if starknet.config.blocks_on_demand
```

Comment:
Since here we only have StateUpdate type we need to check again `blocks_on_demand` mode and I'm not happy about that. Maybe the type system needs to be changed?